Adding time-domain LAL waveforms, replacing functions by classes

bvgoncharov opened the pull request:
I added the following modifications:

1. Now, it is possible to call time-domain LAL waveforms in `modules/waveforms.py`. To facilitate this, function `lal_caller()` is replaced with classes. There is a new basic waveform class, `Waveform(object)`. Time-domain LAL waveforms are either Fourier-transformed into frequency domain in LALSimulation (`class LALFD_Waveform(Waveform)`) or returned in time-domain and Fourier transformed in GWFish (`class LALTD_Waveform(LALFD_Waveform)`) before calculating the Fisher matrix. 
2. In `modules/fishermatrix.py`, derivative and Fisher matrix functions are also replaced by classes.

There are also a few minor cosmetic changes.

Notes:

- Only uniform frequency vectors are supported for time-domain models;

jacopok commented:
Let me say that the things I'm mentioning are minor and overall I think this is a great addition!

janosch314 commented:
I agree that these are good changes, but I don't think that all of Jacopo's comments are minor. For example, the roll-off is not save at it is, and in fact, this is one of the issues we found with time-domain waveforms in the past, which can have important impact on PE. This needs to be tested carefully on signals with varying masses.

bvgoncharov commented:
Thanks for reviewing the code, @janosch314 and @jacopok ! As for the `roll_off` argument in the `fft` function, the function is not actually used anywhere at the moment, and I will remove it to avoid any confusion. The data is conditioned in LAL, so I do not foresee any problems with tapering.

bvgoncharov commented on GWFish/modules/fishermatrix.py:
```diff
@@ -19,80 +22,148 @@ def invertSVD(matrix):
+        self.nd = len(fisher_parameters)
+
+    @property
+    def fm(self):
```
Good point, I added caching and a helper function `update_fm`. I think it might be useful in the future.

bvgoncharov commented on GWFish/modules/fishermatrix.py:
```diff
@@ -19,80 +22,148 @@ def invertSVD(matrix):
-
-            signal1 = det.projection(pv_set1, detector, wave, t_of_f)
-            signal2 = det.projection(pv_set2, detector, wave, t_of_f)
+class Derivative(object):
```
Thanks, I removed inheritance from `object` in the recent commit

bvgoncharov commented on GWFish/modules/fft.py:
```diff
@@ -0,0 +1,42 @@
+
+# The functions below are not used in practice, for now
+
+def fft(hh, dt, t_start, t_end, roll_off = 0.2):
```
This function was not used, so I removed it now. FFT and data conditioning are done entirely in LAL, which should be safe.

bvgoncharov commented on GWFish/modules/fishermatrix.py:
```diff
@@ -19,80 +22,148 @@ def invertSVD(matrix):
+    def projection_at_parameters(self, new_projection_data):
+        self._projection_at_parameters = new_projection_data
+
+    def at_parameter(self, target_parameter):
```
Sounds good, I renamed the function.

bvgoncharov commented on GWFish/modules/fishermatrix.py:
```diff
@@ -19,80 +22,148 @@ def invertSVD(matrix):
+    def __call__(self, target_parameter):
+        return self.at_parameter(target_parameter)
+
+class FisherMatrix(object):
```
Indeed, I think it would be useful to keep it as a class, to more easily add other methods in the future. For example, calculating the inverse. Some parts from `analyzeFisherErrors` can potentially be moved there too.

bvgoncharov commented:
Ok, I think I addressed all the comments, @jacopok and @janosch314 . Unless there are any objections from you and also @u-dupletsa , I could merge this PR within a few days.

As a bonus, I now added citation information to `README.md`, you can see how it looks here: https://github.com/bvgoncharov/GWFish . I think it looks good, but let me know if you would like to change/remove it.

u-dupletsa commented:
Hi Boris, 

Thank you for all the work! 

I’m going through the changes in the code and I don’t know if this is important or not, but I have a question about the reference frequency in the Waveform class. Before we were setting it to 50Hz (with this choice, if I remember well, the LAL TaylorF2 and the GWFish TaylorF2 were almost identical) and now it is set by default to the minimum frequency. I recall that setting f_ref to a low frequency was creating some problems. Is there a reason behind the choice to choose the minimum frequency?

Ulyana

> On 8 Feb 2023, at 12:00, Boris Goncharov ***@***.***> wrote:
> 
> 
> Ok, I think I addressed all the comments, @jacopok <https://github.com/jacopok> and @janosch314 <https://github.com/janosch314> . Unless there are any objections from you and also @u-dupletsa <https://github.com/u-dupletsa> , I could merge this PR within a few days.
> 
> As a bonus, I now added citation information to README.md, you can see how it looks here: https://github.com/bvgoncharov/GWFish <https://github.com/bvgoncharov/GWFish> . I think it looks good, but let me know if you would like to change/remove it.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/janosch314/GWFish/pull/55#issuecomment-1422411515>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ATUPTCDWM7C2DWC2AGNKA7LWWN4EVANCNFSM6AAAAAAUS72SWY>.
> You are receiving this because you were mentioned.
>

bvgoncharov commented:
Hi Ulyana, thanks for looking into the changes! Good point identifying the change in reference frequency, I forgot about it. I introduced it when debugging time-domain waveforms. I considered the following when setting `f_ref = f_low`:

- As discussed [here](https://git.ligo.org/lscsoft/bilby/-/issues/443), some time-domain waveforms require `f_ref = f_low`.
- You can also see in LAL docs for SimInspiralTD [here](https://lscsoft.docs.ligo.org/lalsuite/lalsimulation/group___l_a_l_sim_inspiral__c.html): "If calling with precessing time-domain approximants for which the reference frequency is the starting frequency, or if calling with NR_hdf5 approximant, the starting frequency is not altered".
- Reference frequency was set to low frequency in our PBH work. 
- From what I remember, the waveform looked smoother with `f_ref = f_low`, which I expected could introduce less trouble when FFT'ing time-domain waveforms. Empirically, I ran a quick test now, and I see more consistency between mass parameter errors in PhenomXPHM and PhenomTPHM when setting `f_ref = f_low`, compared to setting e.g. `f_ref = 50 # Hz`.
- I also changed the frequency range a lot and got tired of changing `f_ref` by hand :)

jacopok commented on README.md:
```diff
@@ -1,12 +1,38 @@
 
 Read the documentation [here](https://gwfish.readthedocs.io)!
 
+## Citation
```
Adding the citation is a great idea!
However, I think this is formatted in a more complicated way than it needs: you can just use a code block, like


```
@ARTICLE{DupletsaHarms2023,
        author = {{Dupletsa}, U. and {Harms}, J. and {Banerjee}, B. and {Branchesi}, M. and {Goncharov}, B. and {Maselli}, A. and {Oliveira}, A.~C.~S. and {Ronchini}, S. and {Tissino}, J.},
        title = "{GWFISH: A simulation software to evaluate parameter-estimation capabilities of gravitational-wave detector networks}",
        journal = {Astronomy and Computing},
        keywords = {General Relativity and Quantum Cosmology},
        year = 2023,
        month = jan,
        volume = {42},
        eid = {100671},
        pages = {100671},
        doi = {10.1016/j.ascom.2022.100671},
        archivePrefix = {arXiv},
        eprint = {2205.02499},
        primaryClass = {gr-qc},
        adsurl = {https://ui.adsabs.harvard.edu/abs/2023A&C....4200671D},
        adsnote = {Provided by the SAO/NASA Astrophysics Data System}
}
```

(so, using three backticks: \`\`\` at the beginning and end)

jacopok replied:
and in there you can just have regular spaces, regular newlines and so forth (it's all verbatim rendered)

bvgoncharov replied:
Thanks, done!

jacopok commented on GWFish/modules/waveforms.py:
```diff
@@ -113,51 +130,308 @@ def t_of_f_PN(parameters, frequencyvector):
+
+        polarizations = np.hstack((hfp, hfc))
+
+        print('Warning: inverting hx in LAL caller')
```
I found that when running the [tutorial code](https://gwfish.readthedocs.io/en/latest/tutorials/tutorial_170817.html) this line is spammed on the console.
What does it mean? I don't see any inversion in this function.

bvgoncharov replied:
Yes, that was left by accident, removed

jacopok commented on GWFish/modules/waveforms.py:
```diff
@@ -113,51 +130,308 @@ def t_of_f_PN(parameters, frequencyvector):
+        
+        return polarizations
+
+class LALTD_Waveform(LALFD_Waveform):
```
I think this name is maybe misleading - the waveforms you get when calling this object are still FD, but they come from the TD - maybe `LALFD_Waveform_fromTD`?

bvgoncharov replied:
I see your point, but `FD_fromTD` is also not perfect because the object contains both TD and FD waveform, TD part is not completely discarded. I resolved this by adding more informed docstrings to both classes, they should better explain things.

jacopok commented on GWFish/modules/waveforms.py:
```diff
@@ -113,51 +130,308 @@ def t_of_f_PN(parameters, frequencyvector):
+class LALTD_Waveform(LALFD_Waveform):
+    """ Calls waveforms from LAL in time domain """
+    def __init__(self, name, gw_params, data_params):
+        super(LALTD_Waveform, self).__init__(name, gw_params, data_params)
```
In python3 you can just use `super().__init__(name, gw_params, data_params)`

jacopok commented on GWFish/modules/waveforms.py:
```diff
@@ -113,51 +130,308 @@ def t_of_f_PN(parameters, frequencyvector):
+        self.name = name
+        self._set_default_gw_params()
+        self.gw_params.update(gw_params)
+        self.data_params = data_params
```
Do we really need this layer of abstraction here? if the parameters are just `frequencyvector` and `delta_t` maybe we can just pass them explicitly...

bvgoncharov replied:
I think it would be useful to have more general categories. E.g., for Ulyana's comment I added `f_ref` as another parameter.

bvgoncharov replied:
Also, I was thinking ahead for this removal of `hphc_caller`, and with this in mind, and now it is convenient for all waveform classes to be initiated with the same set of arguments.

jacopok replied:
OK, sounds fair!

jacopok commented on GWFish/modules/waveforms.py:
```diff
@@ -10,29 +11,43 @@
+
+    if time_domain:
+        # Here it is not really t_of_f, it is just a time vector
+        t_of_f = np.arange(0,hphc.shape[0]*waveform_obj.delta_t,waveform_obj.delta_t)
```
This time vector starts from 0, but the user can pass `geocent_time`, the GPS time of merger: this will then be ignored at the moment if the waveform is in the time domain.

Here there should be either a `+parameters['geocent_time']` (and the time array being negative approaching zero), or starting from the frequencyvector (which you have anyway as a property of the `Waveform` object) you can use the `t_of_f_PN` function (less precise, more consistent with the FD waveforms.

bvgoncharov replied:
In fact, I think the original `t_of_f` function is works well here even for time-domain waveforms too. So, I removed this part, only keeping the original code.

jacopok replied:
OK, so if I understand correctly now the `t_of_f` function defined in the base `Waveform` object will be called for all waveforms, be they from TD or FD directly. Seems sensible!

jacopok commented:
Looks good! As written at the moment it breaks the `horizon` module (since it's importing `hphc_amplitudes`), but I think this PR can be merged as-is and I can fix it quickly thereafter.
